overlord/snapstate/check_snap.go

```diff
@@ -81,62 +81,94 @@ func checkSnap(st *state.State, snapFilePath string, si *snap.SideInfo, curInfo
 			return err
 		}
 	}
 
-	switch s.Type {
-	case snap.Type(""), snap.TypeApp, snap.TypeKernel:
-		// "" used in a lot of tests :-/
+	return nil
+}
+
+// CheckSnapCallback defines callbacks for checking a snap for installation or refresh.
+type CheckSnapCallback func(s *state.State, snap, curSnap *snap.Info, flags Flags) error
+
+var checkSnapCallbacks []CheckSnapCallback
+
+// AddCheckSnapCallback installs a callback to check a snap for installation or refresh.
+func AddCheckSnapCallback(check CheckSnapCallback) {
+	checkSnapCallbacks = append(checkSnapCallbacks, check)
+}
+
+func MockCheckSnapCallbacks(checks []CheckSnapCallback) (restore func()) {
+	prev := checkSnapCallbacks
+	checkSnapCallbacks = checks
+	return func() {
+		checkSnapCallbacks = prev
+	}
+}
+
+func checkCoreName(st *state.State, snapInfo, curInfo *snap.Info, flags Flags) error {
+	if snapInfo.Type != snap.TypeOS {
+		// not a relevant check
 		return nil
-	case snap.TypeOS:
-		if curInfo != nil {
-			// already one of these installed
-			return nil
-		}
-		core, err := CoreInfo(st)
-		if err == state.ErrNoState {
-			return nil
-		}
-		if err != nil {
-			return err
-		}
-		if core.Name() != s.Name() {
-			return fmt.Errorf("cannot install core snap %q when core snap %q is already present", s.Name(), core.Name())
-		}
+	}
+	if curInfo != nil {
+		// already one of these installed
+		return nil
+	}
+	core, err := CoreInfo(st)
+	if err == state.ErrNoState {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if core.Name() != snapInfo.Name() {
+		return fmt.Errorf("cannot install core snap %q when core snap %q is already present", snapInfo.Name(), core.Name())
+	}
+
+	return nil
+}
 
+func checkGadget(st *state.State, snapInfo, curInfo *snap.Info, flags Flags) error {
+	if snapInfo.Type != snap.TypeGadget {
+		// not a relevant check
 		return nil
-	case snap.TypeGadget:
-		// gadget specific checks
-		if release.OnClassic {
-			// for the time being
-			return fmt.Errorf("cannot install a gadget snap on classic")
-		}
+	}
+	if release.OnClassic {
+		// for the time being
+		return fmt.Errorf("cannot install a gadget snap on classic")
+	}
 
-		currentGadget, err := GadgetInfo(st)
-		// FIXME: check from the model assertion that its the
-		//        right gadget
-		//
-		// in firstboot we have no gadget yet - that is ok
-		if err == state.ErrNoState {
-			return nil
-		}
-		if err != nil {
-			return fmt.Errorf("cannot find original gadget snap")
-		}
+	currentGadget, err := GadgetInfo(st)
+	// in firstboot we have no gadget yet - that is ok
+	// devicestate considers that case
+	if err == state.ErrNoState {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("cannot find original gadget snap: %v", err)
+	}
 
-		// TODO: actually compare snap ids, from current gadget and candidate
-		if currentGadget.Name() != s.Name() {
-			return fmt.Errorf("cannot replace gadget snap with a different one")
+	if currentGadget.SnapID != "" && snapInfo.SnapID != "" {
```

Comment:
As discussed, we probably want to prevent the case where the snap had an ID from being replaced by one without an id.